### ISSUE TITLE
Similar Issue of Deadlock or Stuck

### ISSUE THREAD

**Issue by monexiao** [ross/requests-futures#60]
Recently, I used requests-futures to do http-get to some web service api in a while loop.
Sometimes, the program is not show any error, but the while loop is stuck and do nothing.
So I found some article about requests + ThreadPoolExecutor

https://github.com/requests/requests/issues/2649
https://stackoverflow.com/questions/40891497/threadpoolexecutor-requests-deadlock

This library is also developed by requests+ThreadPoolExecutor, I would like to know if other people have encountered similar problems using this library. 

If yes, is any others library recommended? Can someone provide suggestions? Thanks a lot.

PS. I used Python 2.7

**Comment by ZeroCool2u** [comment]
@monexiao The docs say "Python >= 3.4 required", so I'd say that's your first problem.

I'm using 3.6 and had no issues.

**Comment by ross** [comment]
I've definitely done this sort of thing in 2.7 as well without problems. Are you using `timeout=N` in your requests? The default is forever and it's pretty common source of problems with requests (not just requests-futures.) `timeout` should really be a required param.

**Comment by monexiao** [comment]
@ross Thanks for your kindly reply. I was thought these are two problems at first.

1. Deadlock: It was resolved in requests library or futures library, or requests-futures?
2.  Timeout: I ignored this, my fault. I had tried it and do not see this error.

This issue (about 1.) as below was close without solution, I still thought 1. was not solved.
https://github.com/requests/requests/issues/2649 

And then, I had try to set timeout in requests-futures and grequests, 2. issue has not happened. Of course, 1. was not happened, and I confirmed 1. was solved via your reminder.
PS. I try grequests because of the reason mentioned above (ny mistake), the grequests uses gevent and I need a alternatively solution.

Finally, I really thank @ross  very much.
